internal/postgres/toddl.go

```diff
@@ -0,0 +1,196 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package postgres
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+	"unicode"
+
+	"github.com/cloudspannerecosystem/harbourbridge/schema"
+	"github.com/cloudspannerecosystem/harbourbridge/spanner/ddl"
+)
+
+// schemaToDDL performs schema conversion from the source DB schema to
+// Spanner. It uses the source schema in conv.srcSchema, and writes
+// the Spanner schema to conv.spSchema.
+func schemaToDDL(conv *Conv) error {
+	for _, srcTable := range conv.srcSchema {
+		spTableName, err := GetSpannerTable(conv, srcTable.Name)
+		if err != nil {
+			return err
+		}
+		var spColNames []string
+		spColDef := make(map[string]ddl.ColumnDef)
+		conv.issues[srcTable.Name] = make(map[string][]schemaIssue)
+		// Iterate over columns using ColNames order.
+		for _, srcColName := range srcTable.ColNames {
+			srcCol := srcTable.ColDef[srcColName]
+			colName, err := GetSpannerCol(conv, srcTable.Name, srcCol.Name, false)
+			spColNames = append(spColNames, colName)
+			if err != nil {
+				return fmt.Errorf("can't get Spanner col: %w", err)
+			}
+			ty, issues := toSpannerType(conv, srcCol.Type.Id, srcCol.Type.Mods)
+			if len(srcCol.Type.ArrayBounds) > 1 {
+				ty = ddl.String{ddl.MaxLength{}}
+				issues = append(issues, multiDimensionalArray)
+			}
+			// TODO: add issues for all elements of srcCol.Ignored.
+			if srcCol.Ignored.ForeignKey {
+				issues = append(issues, foreignKey)
+			}
+			if srcCol.Ignored.Default {
+				issues = append(issues, defaultValue)
+			}
+			if len(issues) > 0 {
+				conv.issues[srcTable.Name][srcCol.Name] = issues
+			}
+			spColDef[colName] = ddl.ColumnDef{
+				Name:    colName,
+				T:       ty,
+				IsArray: len(srcCol.Type.ArrayBounds) == 1,
+				NotNull: srcCol.NotNull,
+				Comment: "From: " + quoteIfNeeded(srcCol.Name) + " " + prType(srcCol.Type),
+			}
+		}
+		comment := "Spanner schema for source table " + quoteIfNeeded(srcTable.Name)
+		conv.spSchema[spTableName] = ddl.CreateTable{spTableName, spColNames, spColDef, cvtPrimaryKeys(conv, srcTable.Name, srcTable.PrimaryKeys), comment}
+	}
+	return nil
+}
+
+// toSpannerType maps a scalar source schema type (defined by id and
+// mods) into a Spanner type. This is the core source-to-Spanner type
+// mapping.  toSpannerType returns the Spanner type and a list of type
+// conversion issues encountered.
+func toSpannerType(conv *Conv, id string, mods []int64) (ddl.ScalarType, []schemaIssue) {
```

Comment:
Currently, these source types are from Postgres. How can we extend this if we want to add MySQL mappings as well? Adding more cases? 

Just curious: what if two types have the same name (type id), but different meanings in two different databases, e.g., Postgres and MySQL?

Reply:
Yes, that is entirely possible.
My guess is that we'll need to know what the source DB is: postgres, mysql, etc.
And then we'll just do the appropriate dispatch based on that and the type. It might make sense to have separate switch statements for each DB.

We might even turn much of the switch into a map (and have a separate map for each src database) - that would also allow us to have users control the map e.g. we configure a default map, but users can override.

For the moment let's just keep the simple structure, and then have it evolve as we add support for other DBs.